cmd/starter/main.go

```diff
@@ -130,6 +130,24 @@ func fetchBuildJob() {
 	}
 	repo := job.Repository
 
+	lock, err := redisClient.Get(fmt.Sprintf("project::%s::build-lock", repo))
+	if err != nil {
+		log.WithField(logrus.Fields{
+			"error": err,
+		}).Error("Unable to fetch lock state")
+		return
+	}
+
+	if lock != nil {
+		// If build-job is currently locked, try again later
+		queueEntry, err := job.ToByte()
+		orFail(err)
+		redisClient.RPush("build-queue", string(queueEntry))
+		return
+	}
+
+	redisClient.Set(fmt.Sprintf("project::%s::build-lock", repo), "locked", 1800, 0, false, false)
```

Comment:
i think there's a TOCTTOU bug here. what happens if:

```
goroutine1: calls redis get, and it completes. (lock == nil)
goroutine2: calls redis get, and it completes. (lock == nil)
goroutine2: calls redis set ("has lock")
goroutine1: calls redis set ("has lock")
```

I think you need one of the redis atomic ops? (not clear to me this handles it)